[GPU Process] [FormControls] Add ControlPart for ImageControlsButton

shallawa opened the pull request:
#### 53cde7bfd6be60b50e8031edf9a150dbe74e30a3
<pre>
[GPU Process] [FormControls] Add ControlPart for ImageControlsButton
<a href="https://bugs.webkit.org/show_bug.cgi?id=250664">https://bugs.webkit.org/show_bug.cgi?id=250664</a>
rdar://104288189

Reviewed by Aditya Keerthi.

Drawing this control part will be moved to GraphicsContext so it can be recorded
in WebProcess and drawn in GPUProcess if DOM rendering is enabled for GPUProcess.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ImageControlsButtonPart.h: Copied from Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::sharedControlFactory):
(WebCore::ControlFactoryMac::servicesRolloverButtonCell const):
(WebCore::ControlFactoryMac::createPlatformImageControlsButton):
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.h: Copied from Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h.
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm: Copied from Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h.
(WebCore::ImageControlsButtonMac::ImageControlsButtonMac):
(WebCore::ImageControlsButtonMac::servicesRolloverButtonCellSize):
(WebCore::ImageControlsButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::imageControlsButtonSize const):
(WebCore::RenderThemeMac::servicesRolloverButtonCell const): Deleted.
(WebCore::RenderThemeMac::paintImageControlsButton): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/259061@main">https://commits.webkit.org/259061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ddf34f6b0b31cc0bf0f68b0ca4216b5cac7cdbe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103788 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36731 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113013 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3799 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96037 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109559 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93806 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25414 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6262 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6431 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12415 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46324 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6230 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8198 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -840,7 +844,7 @@ bool RenderTheme::paint(const RenderBox& box, ControlStates& controlStates, cons
         return paintImageControlsButton(box, paintInfo, integralSnappedRect);
 #endif
     case StyleAppearance::CapsLockIndicator:
-        return paintCapsLockIndicator(box, paintInfo, integralSnappedRect);
```
Let's make this change in a separate PR – there's more cleanup that can be done here.

I've filed https://bugs.webkit.org/show_bug.cgi?id=250725.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm:
```diff
@@ -0,0 +1,67 @@
+
+IntSize ImageControlsButtonMac::servicesRolloverButtonCellSize()
+{
+    auto& controlFactory = static_cast<ControlFactoryMac&>(ControlFactory::sharedControlFactory());
```
Consider adding a `sharedControlFactory` static method on `ControlFactoryMac`, rather than performing the `static_cast` here, similar to `RenderThemeCocoa`.

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
```diff
@@ -50,6 +50,7 @@ class ControlMac : public PlatformControl {
     static void updateFocusedState(NSCell *, const ControlStyle&);
     static void updatePressedState(NSCell *, const ControlStyle&);
 
+    virtual IntSize cellSize() const { return { }; };
```
This looks unused?

shallawa replied:
Removed.

pxlcoder commented on Source/WebCore/rendering/RenderThemeMac.h:
```diff
@@ -187,31 +182,15 @@ class RenderThemeMac final : public RenderThemeCocoa {
-    NSSliderCell *sliderThumbHorizontal() const;
-    NSSliderCell *sliderThumbVertical() const;
-#if ENABLE(DATALIST_ELEMENT)
-    NSCell *listButton() const;
```
I think it would be better to do this cleanup in a separate PR.

shallawa replied:
Done. This is tracked by https://bugs.webkit.org/show_bug.cgi?id=250762

pxlcoder reviewed (commented):
> The renderer for this control part will be created for this markup:
> `<button id="image-controls-button"></button>`

This is not entirely accurate, since the markup needs to be in the shadow subtree of an attachment/image element and `-[WKWebViewConfiguration _imageControlsEnabled]` needs to be true.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
```diff
@@ -30,18 +30,28 @@
 
+#if ENABLE(SERVICE_CONTROLS)
+    NSServicesRolloverButtonCell *servicesRolloverButtonCell() const;
+#endif
```
Nit: This should probably have a newline after it.

shallawa committed:
[GPU Process] [FormControls] Add ControlPart for ImageControlsButton
https://bugs.webkit.org/show_bug.cgi?id=250664
rdar://104288189

Reviewed by Aditya Keerthi.

Drawing this control part will be moved to GraphicsContext so it can be recorded
in WebProcess and drawn in GPUProcess if DOM rendering is enabled for GPUProcess.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ImageControlsButtonPart.h: Copied from Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::sharedControlFactory):
(WebCore::ControlFactoryMac::servicesRolloverButtonCell const):
(WebCore::ControlFactoryMac::createPlatformImageControlsButton):
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.h: Copied from Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h.
* Source/WebCore/platform/graphics/mac/controls/ImageControlsButtonMac.mm: Copied from Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h.
(WebCore::ImageControlsButtonMac::ImageControlsButtonMac):
(WebCore::ImageControlsButtonMac::servicesRolloverButtonCellSize):
(WebCore::ImageControlsButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::imageControlsButtonSize const):
(WebCore::RenderThemeMac::servicesRolloverButtonCell const): Deleted.
(WebCore::RenderThemeMac::paintImageControlsButton): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::decode):

Canonical link: https://commits.webkit.org/259061@main
